xmsannotator/R/compute_scores.R

```diff
@@ -1,35 +1,157 @@
+#'
+#' @import dplyr
+#' @importFrom rlang .data
+compute_isotopes <- function(...,
+                             intensity_deviation_tolerance,
+                             peaks,
+                             mass_defect_tolerance,
+                             rt_tolerance) {
+  query <- tibble(...)
+  isotopic_pattern <- compute_isotopic_pattern(query$molecular_formula)
+  second_most_abundant <- isotopic_pattern$abund[2]
+
+  isotopes <- filter_isotopes(query,
+                              intensity_deviation_tolerance,
+                              peaks,
+                              second_most_abundant,
+                              mass_defect_tolerance,
+                              rt_tolerance)
+  isotopes <- distinct(isotopes)
+  isotopes <- match_isotopes_by_intensity(query,
+                                          isotopes,
+                                          isotopic_pattern,
+                                          intensity_deviation_tolerance)
+}
+
+#' Compute a confidence score for each annotation.
+#'
+#' @param annotation A table with annotated peaks.
+#' @param adduct_weights A weight-by-adduct table.
+#' @param intensity_deviation_tolerance A numeric threshold by which an intensity ratio of two isotopic peaks may differ
+#'  from their actual abundance ratio.
+#' @param mass_defect_tolerance A number. Maximum difference in mass defect between two peaks of the same compound.
+#' @param peaks A peak table containing a peak identifier (unique number), mean intensity, module, and rt cluster
+#'  of each identified peak.
+#' @param rt_tolerance A number. Maximum rt difference for two peaks of the same substance.
+#'
 #' @import dplyr
+#' @import purrr
 #' @importFrom rlang .data
-compute_scores <- function(annotation, adduct_weights, time_tolerance) {
+compute_scores <- function(annotation,
+                           adduct_weights,
+                           intensity_deviation_tolerance = 0.1,
+                           mass_defect_tolerance = 0,
+                           peaks,
+                           rt_tolerance) {
   annotation <- filter(annotation, forms_valid_adduct_pair(.data$molecular_formula, .data$adduct))
-  annotation <- left_join(annotation, adduct_weights, by = "adduct")
 
   isotopes <- semi_join(annotation, adduct_weights, by = "adduct")
-  isotopes <- purrr::pmap_dfr(isotopes, ~tibble())
-  isotopes <- distinct(isotopes)
+  # This can be parallelized on `group_split(group_by(isotopes, molecular_formula))`
+  isotopes <- purrr::pmap_dfr(isotopes,
+                              ~compute_isotopes(..., peaks = peaks,
+                                                rt_tolerance = rt_tolerance,
+                                                intensity_deviation_tolerance = intensity_deviation_tolerance,
+                                                mass_defect_tolerance = mass_defect_tolerance))
 
   annotation <- bind_rows(annotation, isotopes)

```

Comment:
This part should be actually an individual function (as discussed) which adds the isotope information to the annotation table.

The score computation can then be fed this actual table.

Reply:
Moved the isotope detection out of score computation: b2a5c6c6a8ebaae2ade73c2ab17f88d3003877ce and c3bc230e362406a8856b79f522e986ef2b502d7c